program fgethex_test;
#include( "stdlib.hhf" )

var
	f			:file;
	inputL		:lword;
	
	procedure cmp128( left:int128; right:int128 ); @returns( "@ne" );
	begin cmp128;
	
		cld();
		lea( esi, left );
		lea( edi, right );
		mov( 4, ecx );
		repe.cmpsd();
		
	end cmp128;
	
	
begin fgethex_test;

	// Testing fileio.geth128
	
	f.create();
	try
	
		f.open( "fc_geth128.data", fileio.r );
		f.geth128( inputL );
		if( cmp128( inputL, 1 ) ) then
		
			raise($2001);
		
		endif;	
		
		f.geth128( inputL );
		if( cmp128( inputL, 2 ) ) then
		
			raise($2002);
		
		endif;	
		
		f.geth128( inputL );
		if( cmp128( inputL, $FFFF_FFFF_FFFF_FFFF_FFFF_FFFF_FFFF_FFFF ) ) then
		
			raise($2003);
		
		endif;	
		
		f.geth128( inputL );
		if( cmp128( inputL, 0 ) ) then
		
			raise($2004);
		
		endif;
		
		try
		
			f.geth128( inputL );

		  unprotected
		  
		  	// We we got to this point, we failed to raise a conversion error.
			
		  	raise($2005 );
			
		  exception( ex.ConversionError )
		  
		  	// Okay, we succeeded if we got this exception.
			
		  anyexception
		  
		  	// Anything else? Pass it on.
			
		  	raise($2006 );
			
		endtry;
		f.readLn( );	// Clean up after ourselves.
		  		
		
		try
		
			f.geth128( inputL );

		  unprotected
		  
		  	// If we got to this point, we failed to raise an overflow error.
			
		  	raise($2007 );
			
		  exception( ex.ValueOutOfRange )
		  
		  	// Okay, we succeeded if we got this exception.
			
		  anyexception
		  
		  	// Anything else? Pass it on.
			
		  	raise($2008 );
			
		endtry;
		f.readLn( );
		
		
		// Better be at end of file.
		
		if( !f.eof( )) then
		
			raise($2000);
			
		endif;
		stderr.put( "f.geth128 succeeded!" nl );
		
	  anyexception
	  
		stderr.put
		( 
			nl nl 
			"***************************************************" nl
			"f.geth128 failed! eax=", eax, nl
			"Data read did not match data expected in file" nl
			"***************************************************" nl
			nl 
		);
		os.exitProcess(1);
		
	endtry;





	// Testing f.geth64
	
	try
	
		f.open( "fc_geth64.data", fileio.r );

		f.geth64( );
		if( eax <> 1 && edx <> 0 ) then
		
			raise($1001);
		
		endif;	
		
		f.geth64( );
		if( eax <> 2 && edx <> 0 ) then
		
			raise($1002);
		
		endif;	
		
		f.geth64( );
		if( eax<> $30 && edx <> 0 ) then
		
			raise($1003);
		
		endif;	
		
		f.geth64( );
		if( eax<> $4a && edx <> 0 ) then
		
			raise($1004);
		
		endif;
		
		try
		
			f.geth64( );

		  unprotected
		  
		  	// We we got to this point, we failed to raise a conversion error.
			
		  	raise($1005 );
			
		  exception( ex.ConversionError )
		  
		  	// Okay, we succeeded if we got this exception.
			
		  anyexception
		  
		  	// Anything else? Pass it on.
			
		  	raise($1006 );
			
		endtry;
		f.readLn( );	// Clean up after ourselves.
		  		
		
		try
		
			f.geth64( );

		  unprotected
		  
		  	// We we got to this point, we failed to raise an overflow error.
			
		  	raise($1007 );
			
		  exception( ex.ValueOutOfRange )
		  
		  	// Okay, we succeeded if we got this exception.
			
		  anyexception
		  
		  	// Anything else? Pass it on.
			
		  	raise($1008 );
			
		endtry;
		f.readLn( );
		
		
		// Better be at end of file.
		
		if( !f.eof( )) then
		
			raise($1000);
			
		endif;
		stderr.put( "f.geth64 succeeded!" nl );
		
	  anyexception
	  
		stderr.put
		( 
			nl nl 
			"***************************************************" nl
			"f.geth64 failed! eax=", eax, nl
			"Data read did not match data expected in file" nl
			"***************************************************" nl
			nl 
		);
		os.exitProcess(1);
		
	endtry;





	// Testing f.geth32
	
	try
	
		f.open( "fc_geth32.data", fileio.r );

		f.geth32( );
		if( eax <> $A ) then
		
			raise($1011);
		
		endif;	
		
		f.geth32( );
		if( eax <> $B ) then
		
			raise($1012);
		
		endif;	
		
		f.geth32( );
		if( eax <> $C0 ) then
		
			raise($1013);
		
		endif;	
		
		f.geth32( );
		if( eax <> -1 ) then
		
			raise($1014);
		
		endif;
		
		try
		
			f.geth32( );

		  unprotected
		  
		  	// We we got to this point, we failed to raise an overflow error.
			
		  	raise($1015 );
			
		  exception( ex.ValueOutOfRange )
		  
		  	// Okay, we succeeded if we got this exception.
			
		  anyexception
		  
		  	// Anything else? Pass it on.
			
		  	raise($1016 );
			
		endtry;
		f.readLn( );
		
		
		// Better be at end of file.
		
		if( !f.eof( )) then
		
			raise($1010);
			
		endif;
		stderr.put( "f.geth32 succeeded!" nl );
		
	  anyexception
	  
		stderr.put
		( 
			nl nl 
			"***************************************************" nl
			"f.geth32 failed! eax=", eax, nl
			"Data read did not match data expected in file" nl
			"***************************************************" nl
			nl 
		);
		os.exitProcess(1);
		
	endtry;








	// Testing f.geth16
	
	try
	
		f.open( "fc_geth16.data", fileio.r );

		f.geth16( );
		if( eax <> $1000 ) then
		
			raise($1021);
		
		endif;	
		
		f.geth16( );
		if( eax <> $FFFF ) then
		
			raise($1022);
		
		endif;	
		
		f.geth16( );
		if( eax <> $0C ) then
		
			raise($1023);
		
		endif;	
		
		f.geth16( );
		if( eax <> $ff ) then
		
			raise($1024);
		
		endif;
		
		try
		
			f.geth16( );

		  unprotected
		  
		  	// We we got to this point, we failed to raise an overflow error.
			
		  	raise($1025 );
			
		  exception( ex.ValueOutOfRange )
		  
		  	// Okay, we succeeded if we got this exception.
			
		  anyexception
		  
		  	// Anything else? Pass it on.
			
		  	raise($1026 );
			
		endtry;
		f.readLn( );
		
		
		// Better be at end of file.
		
		if( !f.eof( )) then
		
			raise($1020);
			
		endif;
		stderr.put( "f.geth16 succeeded!" nl );
		
	  anyexception
	  
		stderr.put
		( 
			nl nl 
			"***************************************************" nl
			"f.geth16 failed! eax=", eax, nl
			"Data read did not match data expected in file" nl
			"***************************************************" nl
			nl 
		);
		os.exitProcess(1);
		
	endtry;





	// Testing f.geth8
	
	try
	
		f.open( "fc_geth8.data", fileio.r );

		f.geth8( );
		if( eax <> $b ) then
		
			raise($1031);
		
		endif;	
		
		f.geth8( );
		if( eax <> $bb ) then
		
			raise($1032);
		
		endif;	
		
		f.geth8( );
		if( eax <> $0 ) then
		
			raise($1033);
		
		endif;	
		
		f.geth8( );
		if( eax <> $ff ) then
		
			raise($1034);
		
		endif;
		
		try
		
			f.geth8( );

		  unprotected
		  
		  	// We we got to this point, we failed to raise an overflow error.
			
		  	raise($1035 );
			
		  exception( ex.ValueOutOfRange )
		  
		  	// Okay, we succeeded if we got this exception.
			
		  anyexception
		  
		  	// Anything else? Pass it on.
			
		  	raise($1036 );
			
		endtry;
		f.readLn( );
		
		
		// Better be at end of file.
		
		if( !f.eof( )) then
		
			raise($1030);
			
		endif;
		stderr.put( "f.geth8 succeeded!" nl );
		
	  anyexception
	  
		stderr.put
		( 
			nl nl 
			"***************************************************" nl
			"f.geth8 failed! eax=", eax, nl
			"Data read did not match data expected in file" nl
			"***************************************************" nl
			nl 
		);
		os.exitProcess(1);
		
	endtry;




			
end fgethex_test;

													